hamilton/function_modifiers/expanders.py

```diff
@@ -573,29 +600,56 @@ def expand_node(
         fn = node_.callable
         base_doc = node_.documentation
 
-        def dict_generator(*args, **kwargs):
-            dict_generated = fn(*args, **kwargs)
-            if self.fill_with is not None:
-                for field in self.fields:
-                    if field not in dict_generated:
-                        dict_generated[field] = self.fill_with
-            return dict_generated
+        # if fn is async
+        if inspect.iscoroutinefunction(fn):
+
+            async def dict_generator(*args, **kwargs):
+                dict_generated = await fn(*args, **kwargs)
+                if self.fill_with is not None:
+                    for field in self.fields:
+                        if field not in dict_generated:
+                            dict_generated[field] = self.fill_with
+                return dict_generated
+
+        else:
+
+            def dict_generator(*args, **kwargs):
+                dict_generated = fn(*args, **kwargs)
+                if self.fill_with is not None:
+                    for field in self.fields:
+                        if field not in dict_generated:
+                            dict_generated[field] = self.fill_with
+                return dict_generated
 
         output_nodes = [node_.copy_with(callabl=dict_generator)]
 
         for field, field_type in self.fields.items():
             doc_string = base_doc  # default doc string of base function.
 
-            def extractor_fn(
-                field_to_extract: str = field, **kwargs
-            ) -> field_type:  # avoiding problems with closures
-                dt = kwargs[node_.name]
-                if field_to_extract not in dt:
-                    raise base.InvalidDecoratorException(
-                        f"No such field: {field_to_extract} produced by {node_.name}. "
-                        f"It only produced {list(dt.keys())}"
-                    )
-                return kwargs[node_.name][field_to_extract]
+            # if fn is async
+            if inspect.iscoroutinefunction(fn):
+
+                async def extractor_fn(field_to_extract: str = field, **kwargs) -> field_type:
```

Comment:
Why does this need to be `async`? IMO the extractor function *shouldn't* be `async`, and I think we can mix/match. There's no await, so its misleading to make it async...